### DISCUSSION TITLE
CylinderSegment speed

### DISCUSSION THREAD

**Discussion by martinusus** [magpylib/magpylib#581] [Q&A] [answered]
Using the **CylinderSegment** shape with not full circle section angles I have noticed a significant drop in computational speed and increase in RAM consumption when evaluating the B field (compared to when full circle is evaluated).
The evaluation is **42 times slower**. RAM consumption rises from 1.5GB to 6GB for the example below and significantly more when more pixel points are evaluated.

Is that normal/expected? 

In the example below the magnet_Full is a full torus (phi1=0, phi2=360) and magnet_Segment is the same except it is just a section **(phi1=15, phi2=75)**:

```
import numpy as np
import magpylib as magpy
import time

# specification common to both magnets
mag=1300.0
r=5.0 # inner radius
R=10.0 # outer radius
H=3.0 # height

magnet_Full=magpy.magnet.CylinderSegment(magnetization=(0,0,mag), dimension=(r,R,H,0,360), position=[0,0,0])
# the following magnet will be evaluated 42 times slower than the one above
magnet_Segment=magpy.magnet.CylinderSegment(magnetization=(0,0,mag), dimension=(r,R,H,15,75), position=[0,0,0])

N = 1000000 # pixel samples
sensor = magpy.Sensor(position=(0,0,0), pixel=np.full((N,3),2.2))

time_start = time.time()
B_Full = sensor.getB(magnet_Full, sumup=True)
time_Full = time.time() - time_start
print("time spent is", time_Full, 's')

time_start = time.time()
B_Segment = sensor.getB(magnet_Segment, sumup=True)
time_Segment = time.time() - time_start
print("time spent is", time_Segment, 's')

print("Time Segment/Full ratio", time_Segment/time_Full, 's')
```

**Comment by OrtnerMichael** [comment, marked as answer]
Hi @martinusus 

In the case of the full cylinder the computation makes use of the core function `magpy.core.magnet_cylinder_field()` and uses a cut-out operation as described [here](https://magpylib.readthedocs.io/en/latest/examples/examples_01_complex_forms.html#cut-out-operation). We are doing this because, as you have noticed, the full-cylinder computation is much faster (~1-5 us) than the actual cylinder segment computation `magpy.core.magnet_cylinder_segment_field()` which takes ~ 50-100 us.

The main reason behind this is, that the segement field relies on the incomplete elliptic integrals, while the full cylinder field makes use only of complete elliptics or cel functions, which are much faster to compute.

While we plan to have a closer look on the cylinder segment computation in terms of performance this will take quite a while. You could also try to make use of the upcoming `TriangularMesh()` magnet class which requires ~ 1us per Triangle facet. It's already implemented on the [facet_body branch](https://github.com/magpylib/magpylib/tree/trimesh_body), but is still work in progress.

I hope this helps !

sincerely,
Michael

**Reply by martinusus** [reply]
Hi Michael,
thank you, that explains it very well. I had a feeling the complexity of the shape is the culprit. Will stick with the segment (fine discretization of that shape would likely be more demanding), will just limit my parallelization so that the RAM requirement does not go through the roof.
Very nice and unique piece of code and in general a time saver.

Best Regards,

Martin

**Reply by OrtnerMichael** [reply]
@martinusus 

Ty - glad it's helpful !

Please don't forget to set the discussion on "answered".

sincerely,
Michael
